compiler/src/iree/compiler/Codegen/Dialect/GPU/Transforms/CombineBarrierRegions.cpp

```diff
@@ -0,0 +1,133 @@
+};
+
+/// Given two barriers, barrierA and barrierB where A is the op immediately
+/// before B in the block, combine them into a single barrier.
+static LogicalResult
+combineBarrierRegionPair(RewriterBase &rewriter,
+                         IREE::GPU::BarrierRegionOp barrierA,
+                         IREE::GPU::BarrierRegionOp barrierB) {
+  // barrierB must immediately follow barrierA.
+  assert(barrierA->getBlock() == barrierB->getBlock());
+  if (barrierA->getNextNode() != barrierB) {
+    return failure();
+  }
+
+  // Fail if barrierA is used by barrierB, either directly or by implicit
+  // capture.
+  for (auto user : barrierA->getUsers()) {
+    if (user == barrierB || barrierB->isProperAncestor(user)) {
+      return failure();
+    }
+  }
+
+  Location fusedLoc =
+      rewriter.getFusedLoc({barrierA.getLoc(), barrierB.getLoc()});
+
+  // Get the combined operands, result types, and yielded values.
+  SmallVector<Value> combinedOperands = barrierA.getInputs();
+  combinedOperands.append(barrierB.getInputs().begin(),
+                          barrierB.getInputs().end());
+  SmallVector<Type> combinedTypes(barrierA->getResultTypes());
+  combinedTypes.append(barrierB->getResultTypes().begin(),
+                       barrierB->getResultTypes().end());
+
+  auto aYield = cast<IREE::GPU::YieldOp>(barrierA.getBody()->getTerminator());
+  auto bYield = cast<IREE::GPU::YieldOp>(barrierB.getBody()->getTerminator());
+  SmallVector<Value> combinedYields = aYield.getValues();
+  combinedYields.append(bYield.getValues().begin(), bYield.getValues().end());
+
+  // Create the new barrier op.
+  auto combinedBarrierOp = rewriter.create<IREE::GPU::BarrierRegionOp>(
+      fusedLoc, combinedTypes, combinedOperands);
+
+  MutableArrayRef<BlockArgument> barrierABbArgReplacements =
+      combinedBarrierOp.getBody()->getArguments().take_front(
+          barrierA->getNumOperands());
+  MutableArrayRef<BlockArgument> barrierBBbArgReplacements =
+      combinedBarrierOp.getBody()->getArguments().take_back(
+          barrierB->getNumOperands());
+
+  // Merge the bodies of the old barriers into the new one.
+  rewriter.mergeBlocks(barrierA.getBody(), combinedBarrierOp.getBody(),
+                       barrierABbArgReplacements);
+  rewriter.mergeBlocks(barrierB.getBody(), combinedBarrierOp.getBody(),
+                       barrierBBbArgReplacements);
+
+  // Erase the old terminators and create a new one with a concatenated list of
+  // values.
+  rewriter.eraseOp(aYield);
+  rewriter.eraseOp(bYield);
+
+  rewriter.setInsertionPointToEnd(combinedBarrierOp.getBody());
+  rewriter.create<IREE::GPU::YieldOp>(fusedLoc, combinedYields);
+
+  SmallVector<Value> valuesToReplace = barrierA.getResults();
+  ValueRange bResults = barrierB.getResults();
+  valuesToReplace.append(bResults.begin(), bResults.end());
+  rewriter.replaceAllUsesWith(valuesToReplace, combinedBarrierOp.getResults());
+  rewriter.eraseOp(barrierA);
+  rewriter.eraseOp(barrierB);
```

Comment:
Is this safe to erase both ops? I thought it could still be on the pattern rewriter worklist.

Reply:
Yes, this is fine. It will notify the rewriter that the op was erased, removing it from the worklist.